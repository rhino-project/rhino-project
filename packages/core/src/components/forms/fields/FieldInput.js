import PropTypes from 'prop-types';
import { Input } from 'reactstrap';

import {
  useFieldError,
  useFieldRegister,
  useFieldInheritedProps
} from 'rhino/hooks/form';
import { useGlobalComponent } from 'rhino/hooks/overrides';

// FIXME: Is there a better way to handle setValueAs?
export const FieldInputBase = ({ setValueAs, ...props }) => {
  const { extractedProps, inheritedProps } = useFieldInheritedProps(props);
  const { path } = props;

  // https://github.com/react-hook-form/react-hook-form/issues/4414
  // FIXME: All registration props could be extracted in useFieldInheritedProps
  const { ref, ...fieldProps } = useFieldRegister(path, { setValueAs });
  // FIXME: I should be able to get the error from the fieldProps/form context
  const error = useFieldError(path);

  return (
    <Input
      {...extractedProps}
      {...fieldProps}
      innerRef={ref}
      autoComplete="off"
      invalid={!!error}
      {...inheritedProps}
    />
  );
};

FieldInputBase.propTypes = {
  path: PropTypes.string.isRequired
};

const FieldInput = (props) =>
  useGlobalComponent('FieldInput', FieldInputBase, props);

export default FieldInput;
