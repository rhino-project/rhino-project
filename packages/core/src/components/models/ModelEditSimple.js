import PropTypes from 'prop-types';

import { ModelEditProvider } from './ModelEditProvider';
import { useModelEditController } from '../../hooks/controllers';
import { useMemo } from 'react';
import { Spinner } from 'reactstrap';

export const ModelEditSimple = ({ children, fallback = true, ...props }) => {
  const controller = useModelEditController(props);
  const {
    show: { isInitialLoading }
  } = controller;

  // Fallback mirrors React 18 Suspense
  const renderFallback = useMemo(() => {
    if (!isInitialLoading || !fallback) return children;

    if (fallback === true) return <Spinner />;

    return fallback;
  }, [children, fallback, isInitialLoading]);

  return (
    <ModelEditProvider {...controller}>{renderFallback}</ModelEditProvider>
  );
};

ModelEditSimple.propTypes = {
  children: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.node),
    PropTypes.node
  ])
};
