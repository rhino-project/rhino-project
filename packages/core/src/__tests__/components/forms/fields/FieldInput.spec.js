import { QueryClient, QueryClientProvider } from '@tanstack/react-query';
import { render } from '@testing-library/react';
import api from '__tests__/shared/modelFixtures';
import FieldInput from 'rhino/components/forms/fields/FieldInput';
import ModelEditSimple from 'rhino/components/models/ModelEditSimple';
import modelLoader from 'rhino/models';

vi.spyOn(modelLoader, 'api', 'get').mockReturnValue(api);

describe('FieldInput', () => {
  const FormWrapper = ({ children }) => {
    const queryClient = new QueryClient({
      defaultOptions: {
        queries: {
          retry: false
        }
      }
    });

    return (
      <QueryClientProvider client={queryClient}>
        <ModelEditSimple fallback={false} model="user">
          {children}
        </ModelEditSimple>
      </QueryClientProvider>
    );
  };

  it('renders inside of model context', () => {
    const { asFragment } = render(<FieldInput path="dummy" />, {
      wrapper: FormWrapper
    });

    expect(asFragment()).toMatchSnapshot();
  });
});
