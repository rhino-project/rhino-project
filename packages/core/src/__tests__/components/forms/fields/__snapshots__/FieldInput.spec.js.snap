// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`FieldInput > render with disabled 1`] = `
<DocumentFragment>
  <input
    aria-invalid="false"
    autocomplete="off"
    class="form-control"
    disabled=""
    id="dummy"
    name="dummy"
    path="dummy"
    type="text"
  />
</DocumentFragment>
`;

exports[`FieldInput > render with place holder 1`] = `
<DocumentFragment>
  <input
    aria-invalid="false"
    autocomplete="off"
    class="form-control"
    id="dummy"
    name="dummy"
    path="dummy"
    placeholder="placeholder"
    type="text"
  />
</DocumentFragment>
`;

exports[`FieldInput > renders inside of model context 1`] = `
<DocumentFragment>
  <input
    aria-invalid="false"
    autocomplete="off"
    class="form-control"
    id="dummy"
    name="dummy"
    path="dummy"
    type="text"
  />
</DocumentFragment>
`;

exports[`FieldInput > should render with global override shorthand 1`] = `
<DocumentFragment>
  <div>
    Bar
  </div>
</DocumentFragment>
`;
